src/java/org/apache/cassandra/index/sai/disk/v2/V2VectorIndexSearcher.java

```diff
@@ -336,82 +342,122 @@ private SparseFixedBitSet bitSetForSearch()
         {
-            var maxSegmentRowId = metadata.toSegmentRowId(metadata.maxSSTableRowId);
-
-            // if we are brute forcing, we want to build a list of segment row ids, but if not,
-            // we want to build a bitset of ordinals corresponding to the rows
-            SparseFixedBitSet bits = null;
-            IntArrayList rowIds = null;
-            if (numRows <= maxBruteForceRows)
-                rowIds = new IntArrayList();
-            else
-                bits = bitSetForSearch();
-
-            try (var ordinalsView = graph.getOrdinalsView())
+            var comparisonsSavedByBsearch = new Histogram(new SlidingWindowReservoir(10));
+            boolean preferSeqScanToBsearch = false;
+            for (int i = 0; i < keysInRange.size(); i++)
             {
-                for (PrimaryKey primaryKey : keysInRange)
+                var primaryKey = keysInRange.get(i);
+                long sstableRowId = primaryKeyMap.exactRowIdOrCeiling(primaryKey);
+
+                // The current primary key is not in this sstable. Use ceiling to search for the row id
+                // of the next closest primary key in this sstable and skip to that primary key.
+                if (sstableRowId < 0)
                 {
-                    long sstableRowId = primaryKeyMap.exactRowIdForPrimaryKey(primaryKey);
-                    // skip rows that are not in our segment (or more preciesely, have no vectors that were indexed)
-                    // or are not in this segment (exactRowIdForPrimaryKey returns a negative value for not found)
-                    if (sstableRowId < metadata.minSSTableRowId)
-                        continue;
-
-                    // if sstable row id has exceeded current ANN segment, stop
-                    if (sstableRowId > metadata.maxSSTableRowId)
+                    long ceilingRowId = - sstableRowId - 1;
+                    if (ceilingRowId > metadata.maxSSTableRowId)
+                    {
+                        // The next greatest primary key is greater than all the primary keys in this sstable
                         break;
+                    }
+                    var ceilingPrimaryKey = primaryKeyMap.primaryKeyFromRowId(ceilingRowId);
 
-                    int segmentRowId = metadata.toSegmentRowId(sstableRowId);
-                    if (numRows <= maxBruteForceRows)
+                    // adaptively choose either seq scan or bsearch to skip ahead in keysInRange until
+                    // we find one at least as large as the ceiling key
+                    if (preferSeqScanToBsearch)
                     {
-                        rowIds.add(segmentRowId);
+                        int j = 0;
+                        for ( ; i + j < keysInRange.size(); j++)
+                        {
+                            var nextPrimaryKey = primaryKeyMap.primaryKeyFromRowId(j);
+                            if (nextPrimaryKey.compareTo(ceilingPrimaryKey) >= 0)
```

Comment:
Unless I am missing something here, we are comparing primary keys from the same primary key map. If this is the case, we could just compare the row ids. This would save loading the `ceilingPrimaryKey` and `nextPrimaryKey`. The memory allocations for the last jfr posted show a huge number of primary key allocations which seem to be coming from here.

Reply:
In the SAI + ANN processing,

(1) create a RangeIterator (iterator over PKs) across all sstables for each predicate
(2) create a RangeIntersectionIterator across all the RI from (1), to see what rows match all of the AND predicates.
(3) For each sstable, look up which keys in the RII are present in that sstable, and turn those into graph ordinals
(4) perform an ANN search limited to those ordinals

this is step 3, so keysInRange contains keys from all sstables, not just this one.

(that's why everyone started complaining when we switched from the perform-intersections-by-sstable logic -- it was broken, but it was fast, b/c it did all the work with rowIds.)